### PR TITLE
Add converge_if_changed support to resources

### PR DESCRIPTION
**Pull request by bmhughes**
# Description

- Chef 16 is now required
  - Resource partials now in use
- Refactor all HCL resources to use `load_current_value` and `converge_if_changed`
  - Resource notifications now function as per the core resources
  - Changed values are displayed and can be reported upon
  - Server configuration items are written to indiviual files
  - Agent configuration is still accumulated as per previous versions
- Refactor json configuration resource to use `load_current_value` and `converge_if_changed`

## Issues Resolved

- n/a

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by kitchen-porter** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: UPGRADING.md [hunk 1]
```diff
@@ -2,6 +2,33 @@
...
+### Breaking Changes
+
+#### Resource
+
+- The `type` property is now required to be set to allow retrieval by `load_current_value` and will be used to generate the configuration template
+- The `vault_mode` property must be configured for HCL items that are used for both server and agent configuration
+  - Default to :server for items that apply to both server and agent mode
+  - Default to :agent for items that apply only to agent mode
+
+- Agent only configuration resources (auto_auth, template) now require identifiying properties to be specified external to the `config` Hash
```
Comment (ramereth) [inline UPGRADING.md:20]:
```suggestion
- Agent only configuration resources (auto_auth, template) now require identifying properties to be specified external to the `config` Hash
```

File: kitchen.yml [hunk 2]
```diff
@@ -10,10 +10,10 @@ provisioner:
   product_version: <%= ENV['CHEF_VERSION'] || 'latest' %>
   install_strategy: always
   log_level: <%= ENV['CHEF_LOG_LEVEL'] || 'auto' %>
+  multiple_converge: 2
```
Comment (ramereth) [inline kitchen.yml:13]: Should we also add `enforce_idempotency: true` too?

File: metadata.rb [hunk 3]
```diff
@@ -16,4 +16,4 @@
 supports 'suse'
 supports 'ubuntu'
 
-depends 'ark', '~> 5.0.0'
+depends 'ark', '~> 6.0.0'
```
Comment (ramereth) [inline metadata.rb:19]: Shouldn't we have a looser dep on this?
```suggestion
depends 'ark', '>= 6.0'
```

**Review by ramereth — CHANGES_REQUESTED** [review]

File: kitchen.yml (hunk 2 above)
Comment (bmhughes) [reply kitchen.yml:13]: It was primarily put into test the load current value code didn't barf but I can't see why not as those are one and the same in this case.

File: metadata.rb (hunk 3 above)
Comment (bmhughes) [reply metadata.rb:19]: I go with only pinning to major versions generally so I'm ok with that but I'd rather use the fuzzy match or pin it` >=6 <7`. 

I know it's been a hot topic recently but looking at it like a gem dependency I can't see how leaving it open on the upper band is any better.

File: kitchen.yml (hunk 2 above)
Comment (ramereth) [reply kitchen.yml:13]: I've been trying to add that to ensure we don't run into issues down the road.

File: metadata.rb (hunk 3 above)
Comment (ramereth) [reply metadata.rb:19]: If you want to lock just the major then you could do `~> 6.0`. If you do `6.0.0` you'll force users to use `6.0` in the long run which I don't think is ideal with sous-chefs cookbooks.

**Review by ramereth — APPROVED** [review]

**Comment by kitchen-porter** [comment]
Released as: [6.0.0](https://github.com/sous-chefs/vault/releases/tag/6.0.0)

**Pull request merged**